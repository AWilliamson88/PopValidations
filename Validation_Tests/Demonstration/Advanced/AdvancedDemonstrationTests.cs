using ApprovalTests;
using PopValidations_Tests.Demonstration.Advanced.Validators;
using PopValidations_Tests.TestHelpers;
using System.Threading.Tasks;
using Xunit;

namespace PopValidations_Tests.Demonstration.Advanced;

public class AdvancedDemonstrationTests
{
    [Fact]
    public async Task AdvancedValidator_ValidateWithSingleArtistAlbum_ToJson()
    {
        // Arrange
        var fakeAlbumDetailChecker = new FakeAlbumDetailsChecker();
        var runner = ValidationRunnerHelper.BasicRunnerSetup(new AdvancedAlbumValidator(fakeAlbumDetailChecker));

        // Act
        var results = await runner.Validate(AdvancedTestData.SingleArtist());
        var json = JsonConverter.ToJson(results);

        // Assert
        Approvals.VerifyJson(json);
    }

    [Fact]
    public async Task AdvancedValidator_ValidateWithCollaborationAlbum_ToJson()
    {
        // Arrange
        var fakeAlbumDetailChecker = new FakeAlbumDetailsChecker();
        var runner = ValidationRunnerHelper.BasicRunnerSetup(new AdvancedAlbumValidator(fakeAlbumDetailChecker));

        // Act
        var results = await runner.Validate(AdvancedTestData.Collaboration());
        var json = JsonConverter.ToJson(results);

        // Assert
        Approvals.VerifyJson(json);
    }

    [Fact]
    public void BasicValidator_Describe_ToJson()
    {
        // Arrange
        var fakeAlbumDetailChecker = new FakeAlbumDetailsChecker();
        var runner = ValidationRunnerHelper.BasicRunnerSetup(new AdvancedAlbumValidator(fakeAlbumDetailChecker));

        // Act
        var results = runner.Describe();
        var json = JsonConverter.ToJson(results);

        // Assert
        Approvals.VerifyJson(json);
    }
}

