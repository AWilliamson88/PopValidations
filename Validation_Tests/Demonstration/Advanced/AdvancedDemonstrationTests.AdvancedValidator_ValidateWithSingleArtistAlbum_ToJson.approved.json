{
    "results": [
        {
            "property": "Artist",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Artist",
                    "message": "PopValidations_Tests.Demonstration.Advanced.Validators.DesignedToBePatternMatchedSomething is ensured to be false by adding Something to the end.",
                    "group": null
                }
            ]
        },
        {
            "property": "Songs[0]",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Songs[0]",
                    "message": "AdvancedSong { Artist = Disturbed, TrackNumber = , TrackName = Down With The Sickness, Duration = 2.4, Genre =  } Was not found in list of songs we own the rights to",
                    "group": {
                        "name": "Album constains songs we don't own the copyright to",
                        "description": "Album constains songs we don't own the copyright to",
                        "group": null
                    }
                }
            ]
        },
        {
            "property": "Songs[0].TrackNumber",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Songs[0].TrackNumber",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "Songs[0].TrackName",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Songs[0].TrackName",
                    "message": "Is not equal to 'Definitely Not The Correct Song Name.'",
                    "group": null
                }
            ]
        },
        {
            "property": "Songs[0].Duration",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Songs[0].Duration",
                    "message": "Song must have a negative duration.",
                    "group": null
                }
            ]
        },
        {
            "property": "Songs[1]",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Songs[1]",
                    "message": "AdvancedSong { Artist = Disturbed, TrackNumber = 1, TrackName = Down With The Sickness, Duration = 6, Genre =  } Was not found in list of songs we own the rights to",
                    "group": {
                        "name": "Album constains songs we don't own the copyright to",
                        "description": "Album constains songs we don't own the copyright to",
                        "group": null
                    }
                }
            ]
        },
        {
            "property": "Songs[1].TrackName",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Songs[1].TrackName",
                    "message": "Is not equal to 'Definitely Not The Correct Song Name.'",
                    "group": null
                }
            ]
        },
        {
            "property": "Songs[1].Duration",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Songs[1].Duration",
                    "message": "Song must have a negative duration.",
                    "group": null
                },
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Songs[1].Duration",
                    "message": "Is not equal to '10.1'",
                    "group": {
                        "name": "Song Duration is greater than 5 minutes",
                        "description": "Song Duration is greater than 5 minutes",
                        "group": null
                    }
                }
            ]
        }
    ]
}